openpathsampling/pathmover.py

```diff
@@ -1244,6 +1286,146 @@ def move(self, globalstate):
+        the PathMovers to choose from
+    weights : list of floats
+        the relative weight of each PathMover (does not need to be normalized)
+    """
+
+    def _selector(self, globalstate):
+        if self.weights is None:
+            weights = [1.0] * len(self.movers)
+        else:
+            weights = self.weights
+
+        # this is implemented by setting all weights locally to zero that
+        # correspond to movers that will potentially fail since the required
+        # input ensembles are not present in the globalstate
+
+        present_ensembles = globalstate.ensembles
+
+        for idx, mover in enumerate(self.movers):
+            for ens in mover.input_ensembles:
+                if ens not in present_ensembles:
+                    # ens might be required but is not present
+                    weights[idx] = 0.0
+
+        return weights
+
+class FirstAllowedMover(RandomSelectionMover):
+    """
+    Chooses a first mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in enumerate(self.movers):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
+
+class LastAllowedMover(RandomSelectionMover):
+    """
+    Chooses the last mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in reversed(enumerate(self.movers)):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
```

Comment:
> The idea was to have a clean way to write bootstrapping moves the same way as all the other moves.

This is exactly the idea I disagree with. They're not the same. We're shoehorning them in, and everything we've done like that is a kludge. Furthermore, by listing these moves alongside the others, we hide the from the user the fact that they are fundamentally incompatible with other methods. Bootstrapping is really simple: it isn't a move, it is a simulation approach. You have a separate list of moves for each step; you run them until all moves are accepted.

I think the original approach was the best. Trying to make it into something more general with `PartialAcceptance` was a bad idea. We've rewritten bootstrapping 3 or 4 times, each time adding *more* code that "might" be reused, but when we don't have any actual examples where we reuse it. We keep trying to make it more like a normal move, which hides the fact that it isn't.

If you leave it as is, please find a way to mark moves that can not be expected to have a detailed balance partner. I'm in favor of (shortly after 1.0) rewriting bootstrapping and removing partial acceptance.

> And we might reuse the `FirstAllowed` and `LastAllowed`.

When? One of my concerns is that we keep exploding with more and more classes because we "might" reuse them. We're falling into the trap that is the reason programmers use the expression YAGNI: "ya ain't gonna need it." By making too many overly generic classes, I'm concerned that we risk making it easy to do things that aren't physically correct. These moves will never satisfy detailed balance. I think it is a bad idea to include them alongside the moves that *do* satisfy detailed balance.

Adding in something like this later would be easy if we find an actual use for it. But I think we should aim to keep code slim when possible.

Note that the distinction here is that, like PartialAcceptance, you can not *create* a move that satisfies detailed balance. Other first-or-last (like choose first/last subtrajectory) can be made to satisfy detailed balance by wrapping them in a RandomChoice.  

> These two movers fall into the same category as the `SequentialAllowed` mover that is new and was meant for single replica TIS.

So one move in the category does satisfy detailed balance, and the others don't? Isn't that another potential source of confusion?

> I agree that this introduces a few more Movers that could be confusing to a user, but I assume that these functions will usually just be ignored and are only for advances users relevant.

I agree, but I think we shouldn't let that be an excuse for adding code that isn't needed. We have over 200 classes, around 50 in `pathmover.py` alone (according to `grep "^class"`). The larger that number, the higher the barrier to a user contributing. I think we need to work hard at reducing our classes, and turning anything possible into convenience classes of more general treatments.